crates/common/upload/src/upload.rs

```diff
@@ -0,0 +1,244 @@
+    }
+
+    async fn upload_request(
+        &self,
+        url: &UploadInfo,
+        filename: &Utf8PathBuf,
+    ) -> Result<reqwest::Response, UploadError> {
+        let backoff = self.backoff.clone();
+
+        let operation = || async {
+            let file = File::open(filename)
+                .await
+                .context(format!("Can't open a file {:?}", filename))
+                .map_err(backoff::Error::Permanent)?;
+
+            let file_length = file
+                .metadata()
+                .await
+                .context(format!("Can't read a file {:?} metadata", filename))
+                .map_err(backoff::Error::Permanent)?
+                .len();
+
+            let file_body = Body::wrap_stream(FramedRead::new(file, BytesCodec::new()));
+
+            let mut client = reqwest::Client::new().put(url.url());
+
+            if let Some(Auth::Bearer(token)) = &url.auth {
+                client = client.bearer_auth(token)
+            }
+
+            client = client.header(CONTENT_LENGTH, file_length);
+
+            client
+                .body(file_body)
+                .send()
+                .await
+                .map_err(|err| {
+                    if err.is_builder() || err.is_connect() {
+                        backoff::Error::Permanent(UploadError::Network(err))
+                    } else {
+                        backoff::Error::transient(UploadError::Network(err))
+                    }
+                })?
+                .error_for_status()
+                .map_err(|err| match err.status() {
+                    Some(status_error) if status_error.is_client_error() => {
+                        backoff::Error::Permanent(UploadError::Network(err))
+                    }
+                    _ => backoff::Error::transient(UploadError::Network(err)),
+                })
+        };
+
+        retry_notify(backoff, operation, |err, dur: Duration| {
+            let dur = dur.as_secs();
+            warn!("Temporary failure: {err}. Retrying in {dur}s",)
+        })
+        .await
+    }
+
+    async fn retry(&self, url: &UploadInfo) -> Result<(), UploadError> {
+        loop {
+            info!("Could not resume upload, restarting");
+            match self.upload_request(url, &self.source_filename).await {
+                Ok(_) => break,
+
+                Err(UploadError::Network(err)) => {
+                    warn!("Error while upload a file: {err}.\nRetrying...");
+                    continue;
+                }
+
+                Err(UploadError::Io { context, source }) => {
+                    return Err(UploadError::Io { context, source })
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn filename(&self) -> &Utf8Path {
+        self.source_filename.as_path()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tempfile::tempdir;
+    use tokio::io::AsyncWriteExt;
+
+    #[tokio::test]
+    async fn upload_content_no_auth() {
```

Comment:
You can also create a test `upload_content_with_auth()` based on this test.
The change will be only below:

```rust
let _mock1 = server
    .mock("PUT", "/some_file.txt")
    .with_status(201)
    .match_header("Authorization", mockito::Matcher::Regex(r"Bearer .*".to_string()))
    .create();

let url = UploadInfo::new(&target_url).with_auth(Auth::Bearer("1234".to_string()));

```